doc/api/module.md

```diff
@@ -111,6 +111,17 @@ changes:
 Register a module that exports [hooks][] that customize Node.js module
 resolution and loading behavior. See [Customization hooks][].
 
+> **Warning** When setting up a `MessageChannel` to communicate with hooks,
+> beware that this can lead to a deadlock. For example, you have 2 modules,
+> A and B. "A" is registered first and sets up a message channel, which it uses
+> in its `resolve` hook. After "A" is registered, "B" is registered. Resolving
+> "B"'s specifier will go through "A"'s `resolve` hook, which will try to
+> communicate with a locked thread that is busy trying to register "B"'s hooks.
+> Since registering "B" depends on resolving "B"'s specifier, and resolving
+> "B"'s specifier is blocked by "A"'s communication request that is itself
+> blocked by the pending registration that started the chain, the application
+> becomes deadlocked.
```

Comment:
TLDR: if there are multiple loaders and at least one uses MessageChannel, you will probably footgun

Reply:
Could we keep it as that:
```suggestion
> **Warning** If a `resolve` or `load` is left pending on a response from a
> `MessageChannel`, that can lead to a deadlock as the main thread
> might be "asleep" waiting for a result from the loader thread.
```

Reply:
> To avoid that, always set a timeout when dealing with cross thread communication inside those hooks.

This is never going to be a suitable solution, it would be better not to suggest it.

Either the timeout will be too short, and will cancel requests that would not be deadlocked, or it's too long, and will incur a startup penalty equal to the length of the timeout when they are deadlocked. There's no reliable way to tune it to _only_ cancel when it would be deadlocked, without significant perf penalties.

Furthermore, just canceling the request may actually not be what you want. Consider a transpiler loader that converts TypeScript into JavaScript, but has to talk to a service on the main thread to know how to do that correctly. If any `Module.register()` is called after this loader is registered, and the second loader is written in TypeScript, it's going to deadlock until the timer expires, then... what? Throw an error? Serve TypeScript to v8 uncompiled?

Reply:
Yeah, I’d throw an error. After say one second of idling, it’s time to give up. It’s not a perf penalty, it’s a DX thing, querying the main thread is kind of like making a network call, you should always treat the case when you never get a response to not keep your users waiting indefinitely.